datafusion/datasource/src/memory.rs

```diff
@@ -902,4 +1108,319 @@ mod tests {
+            .repartition_evenly_by_size(target_partitions)?
+            .unwrap();
+        assert_eq!(repartitioned_raw_batches.len(), 2);
+        let [ref p1, ref p2] = repartitioned_raw_batches[..] else {
+            unreachable!()
+        };
+        // p1=batch(100_000)
+        assert_eq!(p1.len(), 1);
+        assert_eq!(p1[0].num_rows(), 100_000);
+        // p2=[batch(10_000), batch(100), batch(1)]
+        assert_eq!(p2.len(), 3);
+        assert_eq!(p2[0].num_rows(), 10_000);
+        assert_eq!(p2[1].num_rows(), 100);
+        assert_eq!(p2[2].num_rows(), 1);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_repartition_no_sort_information_no_output_ordering() -> Result<()> {
+        let no_sort = vec![];
+        let no_output_ordering = None;
+
+        // Test: common set of functionality
+        run_all_test_scenarios(no_output_ordering.clone(), no_sort.clone())?;
+
+        // Test: does not preserve separate partitions (with own internal ordering) on even split
+        let target_partitions = 3;
+        let mem_src_config =
+            memorysrcconfig_2_partition_with_different_sized_batches(no_sort)?;
+        let partitioned_datasrc = mem_src_config.clone().repartitioned(
+            target_partitions,
+            usize::MAX,
+            no_output_ordering,
+        )?;
+        assert_partitioning(partitioned_datasrc.clone(), Some(3));
+        // Starting = batch(100_000), batch(10_000), batch(1_000), batch(2_000), batch(20)
+        // It should have split as p1=batch(100_000), p2=batch(10_000),  p3=rest(mixed across original partitions)
+        let repartitioned_raw_batches = mem_src_config
+            .repartition_evenly_by_size(target_partitions)?
+            .unwrap();
+        assert_eq!(repartitioned_raw_batches.len(), 3);
+        let [ref p1, ref p2, ref p3] = repartitioned_raw_batches[..] else {
+            unreachable!()
+        };
+        // p1=batch(100_000)
+        assert_eq!(p1.len(), 1);
+        assert_eq!(p1[0].num_rows(), 100_000);
+        // p2=batch(10_000)
+        assert_eq!(p2.len(), 1);
+        assert_eq!(p2[0].num_rows(), 10_000);
+        // p3= batch(1_000), batch(2_000), batch(20)
+        assert_eq!(p3.len(), 3);
+        assert_eq!(p3[0].num_rows(), 1_000);
+        assert_eq!(p3[1].num_rows(), 2_000);
+        assert_eq!(p3[2].num_rows(), 20);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_repartition_with_sort_information() -> Result<()> {
+        let schema = schema();
+        let sort_key = LexOrdering::new(vec![PhysicalSortExpr {
+            expr: col("c", &schema).unwrap(),
+            options: SortOptions::default(),
+        }]);
+        let has_sort = vec![sort_key.clone()];
+        let output_ordering = Some(sort_key);
+
+        // Test: common set of functionality
+        run_all_test_scenarios(output_ordering.clone(), has_sort.clone())?;
+
+        // Test: DOES preserve separate partitions (with own internal ordering)
```

Comment:
Here is the preservation of sort ordering. Again, using the [FileSourceConfig implementation](https://github.com/apache/datafusion/pull/15409/files#r2011495377) as a guide.